### PR TITLE
Add multilingual models support

### PR DESCRIPTION
**Pull request by Edresson**
This PR implements support for multilingual models.  I and @WeberJulian  have been working for some time with multilingual models. However, with the introduction of the new training API (a nice addition :)) the pull request #11 became outdated and made Merge impossible. For this reason, I used @WeberJulian 's great  PR #11 as a base and implemented support for the trainer API. Thanks, @WeberJulian  :).

- [x] Multilingual data loader support 
- [x] Multilingual inference support
- [x] Language weighted sampler (to avoid problems with batch unbalance)
- [x] YourTTS training support
- [x] Unit tests
- [ ] Availability of pre-trained models

Contributions are welcome to support multilingual training for other models such as the Tacotron 2.

### CODE REVIEW

**Comment by WeberJulian** [comment]
I'll add support for Tacotron 2, the generated encoder, and the perfect batch sampler this week. Thank you @Edresson for taking the time to make our changes up to date.

File: TTS/tts/models/base_tts.py [hunk 1]
```diff
@@ -199,6 +209,13 @@ def get_data_loader(
             if hasattr(self, "make_symbols"):
                 custom_symbols = self.make_symbols(self.config)
 
+            if hasattr(self, "language_manager"):
```
Comment (erogol) [inline TTS/tts/models/base_tts.py:212]: I think speaker manager can handle the language mapping too

Comment (erogol) [reply TTS/tts/models/base_tts.py:212]: Also, BaseTTS should not include language settings if we don't plan to support multiple language training in all the TTS models. 

It is better to create BaseMultiLangTTS inheriting BaseTTS

File: TTS/tts/models/base_tts.py [hunk 2]
```diff
@@ -219,10 +236,11 @@ def get_data_loader(
                 use_phonemes=config.use_phonemes,
                 phoneme_language=config.phoneme_language,
                 enable_eos_bos=config.enable_eos_bos_chars,
-                use_noise_augment=not is_eval,
+                use_noise_augment=False if is_eval else config.use_noise_augment,
```
Comment (erogol) [inline TTS/tts/models/base_tts.py:239]: Nice catch

File: TTS/tts/models/base_tts.py [hunk 3]
```diff
@@ -267,6 +285,13 @@ def get_data_loader(
 
             # sampler for DDP
             sampler = DistributedSampler(dataset) if num_gpus > 1 else None
+            if sampler is None:
```
Comment (erogol) [inline TTS/tts/models/base_tts.py:288]: This also should live in BaseMultiLingualTTS

File: TTS/tts/models/vits.py [hunk 4]
```diff
@@ -1,20 +1,22 @@
...
 from torch.cuda.amp.autocast_mode import autocast
+from torch.nn import functional as F
 
 from TTS.tts.layers.glow_tts.duration_predictor import DurationPredictor
 from TTS.tts.layers.vits.discriminator import VitsDiscriminator
 from TTS.tts.layers.vits.networks import PosteriorEncoder, ResidualCouplingBlocks, TextEncoder
 from TTS.tts.layers.vits.stochastic_duration_predictor import StochasticDurationPredictor
 from TTS.tts.models.base_tts import BaseTTS
 from TTS.tts.utils.helpers import generate_path, maximum_path, rand_segments, segment, sequence_mask
+from TTS.tts.utils.languages import get_language_manager
```
Comment (erogol) [inline TTS/tts/models/vits.py:19]: If we can use SpeakerManager for languages then we don't need it .

File: TTS/tts/models/vits.py [hunk 5]
```diff
@@ -221,7 +263,7 @@ def __init__(self, config: Coqpit, speaker_manager: SpeakerManager = None):
         super().__init__(config)
 
         self.END2END = True
-
+        self.audio_config = config["audio"]
```
Comment (erogol) [inline TTS/tts/models/vits.py:266]: why do you need this ? There is already self.config

File: TTS/tts/models/vits.py [hunk 6]
```diff
@@ -283,16 +327,22 @@ def __init__(self, config: Coqpit, speaker_manager: SpeakerManager = None):
...
             )
         else:
             self.duration_predictor = DurationPredictor(
-                args.hidden_channels, 256, 3, args.dropout_p_duration_predictor, cond_channels=self.embedded_speaker_dim
+                args.hidden_channels + self.embedded_language_dim,
+                256,
+                3,
+                args.dropout_p_duration_predictor,
+                cond_channels=self.embedded_speaker_dim,
+                language_emb_dim=self.embedded_language_dim,
```
Comment (erogol) [inline TTS/tts/models/vits.py:345]: Why do we pass. `args.hidden_channels + self.embedded_language_dim`, if we already pass                 `language_emb_dim=self.embedded_language_dim`,

File: TTS/tts/models/vits.py [hunk 7]
```diff
@@ -179,11 +208,24 @@ class VitsArgs(Coqpit):
...
+    speaker_encoder_model_path: str = ""
+    fine_tuning_mode: int = 0
+    freeze_encoder: bool = False
+    freeze_DP: bool = False
+    freeze_PE: bool = False
+    freeze_flow_decoder: bool = False
+    freeze_waveform_decoder: bool = False
 
 
 class Vits(BaseTTS):
```
Comment (erogol) [inline TTS/tts/models/vits.py:231]: This is a lot of code change. Can we maybe create a new model VITSMultiLingual and inherit VITS ?

File: TTS/tts/models/vits.py [hunk 8]
```diff
@@ -351,29 +428,111 @@ def _init_d_vector(self, config):
...
+
+        return sid, g, lid
 
     def get_aux_input(self, aux_input: Dict):
-        sid, g = self._set_cond_input(aux_input)
-        return {"speaker_id": sid, "style_wav": None, "d_vector": g}
+        sid, g, lid = self._set_cond_input(aux_input)
+        return {"speaker_id": sid, "style_wav": None, "d_vector": g, "language_id": lid}
+
+    def get_aux_input_from_test_setences(self, sentence_info):
```
Comment (erogol) [inline TTS/tts/models/vits.py:481]: `test_sentences`

File: TTS/tts/models/vits.py [hunk 9]
```diff
@@ -351,29 +428,111 @@ def _init_d_vector(self, config):
...
+            if len(sentence_info) == 1:
+                text = sentence_info[0]
+            elif len(sentence_info) == 2:
+                text, speaker_name = sentence_info
+            elif len(sentence_info) == 3:
+                text, speaker_name, style_wav = sentence_info
+            elif len(sentence_info) == 4:
+                text, speaker_name, style_wav, language_name = sentence_info
+        else:
+            text = sentence_info
```
Comment (erogol) [inline TTS/tts/models/vits.py:500]: To prevent this we can use dict for test sentences

File: TTS/tts/models/vits.py [hunk 10]
```diff
@@ -396,13 +555,18 @@ def forward(
             - speaker_ids: :math:`[B]`
         """
         outputs = {}
-        sid, g = self._set_cond_input(aux_input)
-        x, m_p, logs_p, x_mask = self.text_encoder(x, x_lengths)
-
+        sid, g, lid = self._set_cond_input(aux_input)
         # speaker embedding
-        if self.num_speakers > 1 and sid is not None:
+        if self.args.use_speaker_embedding and sid is not None and not self.use_d_vector:
```
Comment (erogol) [inline TTS/tts/models/vits.py:560]: `self.args.use_speaker_embedding and sid is not None` should be enough ?

File: tests/data_tests/test_loader.py [hunk 11]
```diff
@@ -38,6 +38,11 @@ def __init__(self, *args, **kwargs):
 
     def _create_dataloader(self, batch_size, r, bgs):
         items = ljspeech(c.data_path, "metadata.csv")
+
```
Comment (erogol) [inline tests/data_tests/test_loader.py:41]: Well changes should not effect the existing code

This is monkey patching :)

File: notebooks/VITS_d-vector_multilingual_exemple.ipynb [hunk 12]
```diff
@@ -0,0 +1,223 @@
+{
```
Comment (erogol) [inline notebooks/VITS_d-vector_multilingual_exemple.ipynb:1]: I don't want to host notebooks anymore for model testing. They are terrible to maintain

File: TTS/utils/io.py [hunk 13]
```diff
@@ -38,7 +38,7 @@ def copy_model_files(config: Coqpit, out_path, new_fields):
     """
     copy_config_path = os.path.join(out_path, "config.json")
     # add extra information fields
-    if new_fields:
+    if new_fields is not None:
```
Comment (erogol) [inline TTS/utils/io.py:41]: Maybe set `new_fields=None` by default

File: TTS/tts/utils/synthesis.py [hunk 14]
```diff
@@ -71,6 +71,7 @@ def run_model_torch(
     speaker_id: int = None,
     style_mel: torch.Tensor = None,
     d_vector: torch.Tensor = None,
+    language_id: torch.Tensor = None,
```
Comment (erogol) [inline TTS/tts/utils/synthesis.py:74]: Add to docstring

File: TTS/tts/utils/synthesis.py [hunk 15]
```diff
@@ -160,19 +162,20 @@ def inv_spectrogram(postnet_output, ap, CONFIG):
     return wav
 
 
-def speaker_id_to_torch(speaker_id, cuda=False):
-    if speaker_id is not None:
-        speaker_id = np.asarray(speaker_id)
-        speaker_id = torch.from_numpy(speaker_id)
+def id_to_torch(aux_id, cuda=False):
```
Comment (erogol) [inline TTS/tts/utils/synthesis.py:165]: Why the renaming?

File: TTS/tts/models/base_tts.py [hunk 16]
```diff
@@ -188,8 +191,15 @@ def get_data_loader(
...
             # setup multi-speaker attributes
             if hasattr(self, "speaker_manager") and self.speaker_manager is not None:
-                speaker_id_mapping = self.speaker_manager.speaker_ids if config.use_speaker_embedding else None
-                d_vector_mapping = self.speaker_manager.d_vectors if config.use_d_vector_file else None
+                if hasattr(config, "model_args"):
+                    speaker_id_mapping = (
+                        self.speaker_manager.speaker_ids if config.model_args.use_speaker_embedding else None
+                    )
+                    d_vector_mapping = self.speaker_manager.d_vectors if config.model_args.use_d_vector_file else None
+                    config.use_d_vector_file = config.model_args.use_d_vector_file
```
Comment (erogol) [inline TTS/tts/models/base_tts.py:199]: I don't think this part generalizes to all the models

File: TTS/tts/models/vits.py [hunk 17]
```diff
@@ -179,11 +208,24 @@ class VitsArgs(Coqpit):
...
+    num_languages: int = 0
+    use_speaker_encoder_as_loss: bool = False
+    speaker_encoder_config_path: str = ""
+    speaker_encoder_model_path: str = ""
+    fine_tuning_mode: int = 0
+    freeze_encoder: bool = False
+    freeze_DP: bool = False
+    freeze_PE: bool = False
+    freeze_flow_decoder: bool = False
+    freeze_waveform_decoder: bool = False
```
Comment (erogol) [inline TTS/tts/models/vits.py:228]: Docstring ?

File: TTS/tts/models/vits.py [hunk 18]
```diff
@@ -334,6 +383,34 @@ def init_multispeaker(self, config: Coqpit):
...
+            if (
+                hasattr(self.speaker_encoder, "audio_config")
+                and self.audio_config["sample_rate"] != self.speaker_encoder.audio_config["sample_rate"]
+            ):
+                self.audio_transform = torchaudio.transforms.Resample(
+                    orig_freq=self.audio_config["sample_rate"],
+                    new_freq=self.speaker_encoder.audio_config["sample_rate"],
+                )
+            else:
+                self.audio_transform = None
```
Comment (erogol) [inline TTS/tts/models/vits.py:409]: make this a function and just call it here

File: TTS/tts/models/vits.py [hunk 19]
```diff
@@ -396,13 +555,18 @@ def forward(
...
+        sid, g, lid = self._set_cond_input(aux_input)
         # speaker embedding
-        if self.num_speakers > 1 and sid is not None:
+        if self.args.use_speaker_embedding and sid is not None and not self.use_d_vector:
             g = self.emb_g(sid).unsqueeze(-1)  # [b, h, 1]
 
+        # language embedding
+        lang_emb = None
+        if self.args.use_language_embedding and lid is not None:
+            lang_emb = self.emb_l(lid).unsqueeze(-1)
```
Comment (erogol) [inline TTS/tts/models/vits.py:566]: it is better to  check against `hasattr(self, "emb_l")` to not to miss broken config

File: TTS/tts/models/vits.py [hunk 20]
```diff
@@ -396,13 +555,18 @@ def forward(
             - speaker_ids: :math:`[B]`
         """
         outputs = {}
-        sid, g = self._set_cond_input(aux_input)
-        x, m_p, logs_p, x_mask = self.text_encoder(x, x_lengths)
-
+        sid, g, lid = self._set_cond_input(aux_input)
```
Comment (erogol) [inline TTS/tts/models/vits.py:558]: Update the docstrings

File: TTS/tts/models/vits.py [hunk 21]
```diff
@@ -447,40 +613,189 @@ def forward(
...
+            # resample audio to speaker encoder sample_rate
+            if self.audio_transform is not None:
+                wavs_batch = self.audio_transform(wavs_batch)
+
+            pred_embs = self.speaker_encoder.forward(wavs_batch, l2_norm=True)
+
+            # split generated and GT speaker embeddings
+            gt_spk_emb, syn_spk_emb = torch.chunk(pred_embs, 2, dim=0)
+        else:
+            gt_spk_emb, syn_spk_emb = None, None
```
Comment (erogol) [inline TTS/tts/models/vits.py:636]: Put it into a separate function

File: TTS/tts/models/vits.py [hunk 22]
```diff
@@ -447,40 +613,189 @@ def forward(
...
+    def forward_fine_tuning(
+        self,
+        x: torch.tensor,
+        x_lengths: torch.tensor,
+        y: torch.tensor,
+        y_lengths: torch.tensor,
+        aux_input={"d_vectors": None, "speaker_ids": None, "language_ids": None},
+        waveform=None,
+    ) -> Dict:
+        """Forward pass of the model.
```
Comment (erogol) [inline TTS/tts/models/vits.py:664]: How different is it from the `forward()`

Comment (erogol) [reply TTS/tts/models/vits.py:664]: Maybe a sample code snippet too

File: TTS/tts/models/vits.py [hunk 23]
```diff
@@ -499,12 +814,21 @@ def inference(self, x, aux_input={"d_vectors": None, "speaker_ids": None}):
         outputs = {"model_outputs": o, "alignments": attn.squeeze(1), "z": z, "z_p": z_p, "m_p": m_p, "logs_p": logs_p}
         return outputs
 
-    def voice_conversion(self, y, y_lengths, sid_src, sid_tgt):
+    def voice_conversion(self, y, y_lengths, speaker_cond_src, speaker_cond_tgt):
```
Comment (erogol) [inline TTS/tts/models/vits.py:817]: It's a crazy long function. Maybe split it into smaller functions.

File: TTS/tts/utils/speakers.py [hunk 24]
```diff
@@ -154,15 +155,21 @@ def save_d_vectors_to_file(self, file_path: str) -> None:
         """
         self._save_json(file_path, self.d_vectors)
 
-    def set_d_vectors_from_file(self, file_path: str) -> None:
+    def set_d_vectors_from_file(self, file_path: str, data: List = None) -> None:
```
Comment (erogol) [inline TTS/tts/utils/speakers.py:158]: Function name already expects file_path. It doesn't make sense to add `data`. Then create a new function if it is needed.

File: TTS/tts/utils/speakers.py [hunk 25]
```diff
@@ -154,15 +155,21 @@ def save_d_vectors_to_file(self, file_path: str) -> None:
...
         self.d_vectors = self._load_json(file_path)
-        speakers = sorted({x["name"] for x in self.d_vectors.values()})
-        self.speaker_ids = {name: i for i, name in enumerate(speakers)}
+
+        # load speakers from data, because during the training we can just use some speakers from d_vector_file
+        if data is not None:
+            self.speaker_ids, _ = self.parse_speakers_from_data(data)
+        else:
+            speakers = sorted({x["name"] for x in self.d_vectors.values()})
+            self.speaker_ids = {name: i for i, name in enumerate(speakers)}
```
Comment (erogol) [inline TTS/tts/utils/speakers.py:171]: I don't understand the use-case

File: TTS/tts/utils/speakers.py [hunk 26]
```diff
@@ -187,6 +194,20 @@ def get_d_vectors_by_speaker(self, speaker_idx: str) -> List[List]:
         """
         return [x["embedding"] for x in self.d_vectors.values() if x["name"] == speaker_idx]
 
+    def get_d_vector_by_speaker(self, speaker_idx: str) -> np.ndarray:
```
Comment (erogol) [inline TTS/tts/utils/speakers.py:197]: This returns the first matching d-vector of the speaker right

Comment (erogol) [reply TTS/tts/utils/speakers.py:197]: If so if you set num_samples=1 in the function below it is the same thing

File: TTS/tts/utils/speakers.py [hunk 27]
```diff
@@ -242,12 +289,16 @@ def compute_d_vector_from_clip(self, wav_file: Union[str, list]) -> list:
 
         def _compute(wav_file: str):
             waveform = self.speaker_encoder_ap.load_wav(wav_file, sr=self.speaker_encoder_ap.sample_rate)
-            spec = self.speaker_encoder_ap.melspectrogram(waveform)
-            spec = torch.from_numpy(spec.T)
+            if not self.speaker_encoder_config.model_params.get("use_torch_spec", False):
```
Comment (erogol) [inline TTS/tts/utils/speakers.py:292]: This attribute has nothing to do here. It is again Monkey Patching

Comment (erogol) [reply TTS/tts/utils/speakers.py:292]: Speaker Encoder should decide what to use internally

File: TTS/tts/utils/speakers.py [hunk 28]
```diff
@@ -349,7 +400,7 @@ def get_speaker_manager(c: Coqpit, data: List = None, restore_path: str = None,
                             "You must copy the file speakers.json to restore_path, or set a valid file in CONFIG.d_vector_file"
                         )
                     speaker_manager.load_d_vectors_file(c.d_vector_file)
-                speaker_manager.set_d_vectors_from_file(speakers_file)
```
Comment (erogol) [inline TTS/tts/utils/speakers.py:352]: get_speaker_manager is gonna be removed. User should create its own object as in the multi-speaker VCTK recipes.

**Comment by erogol** [comment]
The most challenging PR I've tried to review ever. Next time please split things into PRs

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/tts/models/vits.py (hunk 10 above)
Comment (WeberJulian) [reply TTS/tts/models/vits.py:560]: Indeed

File: TTS/tts/layers/vits/stochastic_duration_predictor.py [hunk 29]
```diff
@@ -266,7 +280,7 @@ def forward(self, x, x_mask, dr=None, g=None, reverse=False, noise_scale=1.0):
 
         flows = list(reversed(self.flows))
         flows = flows[:-2] + [flows[-1]]  # remove a useless vflow
-        z = torch.rand(x.size(0), 2, x.size(2)).to(device=x.device, dtype=x.dtype) * noise_scale
```
[the remark this answers is not in the document]
Comment (WeberJulian) [reply TTS/tts/layers/vits/stochastic_duration_predictor.py:269]: not so much when you see how small are the differences between different inference runs

File: TTS/speaker_encoder/models/resnet.py [hunk 30]
```diff
@@ -1,10 +1,24 @@
 import numpy as np
 import torch
+import torchaudio
 from torch import nn
 
 from TTS.utils.io import load_fsspec
 
 
+class PreEmphasis(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/speaker_encoder/models/resnet.py:9]: I'm not sure but we need to have the same spectrogram as the original implementation as we are using the checkpoint provided by them. For this reason, I implemented it in this way (and because it is very efficient,  all processing will be realized in the GPU)

File: TTS/speaker_encoder/models/resnet.py [hunk 31]
```diff
@@ -140,9 +174,13 @@ def new_parameter(self, *size):
         return out
 
     def forward(self, x, l2_norm=False):
-        x = x.transpose(1, 2)
         with torch.no_grad():
             with torch.cuda.amp.autocast(enabled=False):
+                if self.use_torch_spec:
+                    x = self.torch_spec(x)
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/speaker_encoder/models/resnet.py:180]: yes, if it is False the Mel spec will be computed using torchaudio

File: TTS/trainer.py [hunk 32]
```diff
@@ -439,7 +455,7 @@ def _restore_list_objs(states, obj):
             if "scaler" in checkpoint and self.use_amp_scaler and checkpoint["scaler"]:
                 print(" > Restoring Scaler...")
                 scaler = _restore_list_objs(checkpoint["scaler"], scaler)
-        except (KeyError, RuntimeError):
+        except (KeyError, RuntimeError, ValueError):
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/trainer.py:458]: Because I found a case where an error occurred and returned ValueError and thus made the partial load of the model unfeasible. I don't remember what the case was anymore (maybe different shapes in the layers during restore?) but I find it necessary to always try to partial load as it is more efficient than training from scratch.

File: TTS/tts/configs/vits_config.py [hunk 33]
```diff
@@ -130,13 +131,13 @@ class VitsConfig(BaseTTSConfig):
     add_blank: bool = True
 
     # testing
-    test_sentences: List[str] = field(
+    test_sentences: List[List] = field(
         default_factory=lambda: [
-            "It took me quite a long time to develop a voice, and now that I have it I'm not going to be silent.",
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/tts/configs/vits_config.py:135]: Because from now on the sentences for testing during training must inform the text, speaker_name, style_wav, language_name. As it was in the previous version it wasn't possible to pass the speaker identity and style_wav, we got a style_wav and speaker embedding randomly. But language is not possible to do this so I proposed this solution now the user must provide for each sentence a list with [text, speaker_name, style_wav, language_name]. However, a list in the formats [text] or [text, speaker_name] or [text, speaker_name, style_wav] is accepted to avoid the user needing to inform items not necessary for the training of the model in question.

File: TTS/tts/datasets/formatters.py [hunk 34]
```diff
@@ -59,46 +59,56 @@ def mozilla_de(root_path, meta_file):
...
-    
+
     Args:
-        root_path (str): root folder of the MAILAB language folder. 
+        root_path (str): root folder of the MAILAB language folder.
         meta_files (str):  list of meta files to be used in the training. If None, finds all the csv files
             recursively. Defaults to None
     """
     speaker_regex = re.compile("by_book/(male|female)/(?P<speaker_name>[^/]+)/")
-    if meta_files is None:
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/tts/datasets/formatters.py:71]: This solves a bug ! In the dataset config the  meta_file_train and meta_file_val are defined as str so when we pass the metadata for this function  meta_files  will be empty str ("") and it will generate a Bug ...

File: TTS/tts/utils/speakers.py (hunk 25 above)
Comment (Edresson) [reply TTS/tts/utils/speakers.py:171]: the `data` is the items from the dataset. We need to print to the user only the speakers used in training and not all those present in d_vectors. With this speaker_ids filter, you will only have the speakers that will actually be used in training, otherwise, it will be confusing for the user because he will see all the speakers present in d_vectors and not actually only those who have samples in the training. And so it would be impossible for the user to know if the data loader's speaker filtering is in fact working.  Do you still think a new function is needed? whenever we call set_d_vectors_from_file we would have to call this function... Why not keep in this function?

File: TTS/tts/utils/synthesis.py (hunk 15 above)
Comment (WeberJulian) [reply TTS/tts/utils/synthesis.py:165]: So we can use it for language_id

File: TTS/tts/utils/speakers.py (hunk 27 above)
Comment (Edresson) [reply TTS/tts/utils/speakers.py:292]: I don't understand what you mean. Given that the speaker encoder doesn't have access to AP. So is here the right place to extract the mel-spectrogram. Do you suggest that we need to pass the AP to the Speaker encoder and extract the mel-spec in the Speaker encoder class?

This "IF" just controls the input. The new speaker encoder needs to receive a wave as input.  To remove this "IF" we need to pass both the mel_spec and waveform to the speaker encoder or pass the AP to the speaker encoder  ...

File: TTS/bin/remove_silence_using_vad.py [hunk 35]
```diff
@@ -0,0 +1,219 @@
...
+import glob
+import multiprocessing
+import os
+import pathlib
+import wave
+import webrtcvad
+from tqdm.contrib.concurrent import process_map
+
+
+def read_wave(path):
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/bin/remove_silence_using_vad.py]: Done

File: TTS/tts/utils/speakers.py (hunk 26 above)
Comment (Edresson) [reply TTS/tts/utils/speakers.py:197]: Done

File: TTS/tts/utils/speakers.py (hunk 25 above)
Comment (Edresson) [reply TTS/tts/utils/speakers.py:171]: Done

File: TTS/tts/utils/speakers.py (hunk 28 above)
Comment (Edresson) [reply TTS/tts/utils/speakers.py]: ok, I removed the call of this function in TTS/bin/extract_tts_spectrograms.py. For now, it is used in BaseTTS, but if you like I can remove it as well.

File: TTS/utils/io.py (hunk 13 above)
Comment (Edresson) [reply TTS/utils/io.py:41]: Done

File: TTS/speaker_encoder/models/resnet.py [hunk 36]
```diff
@@ -88,6 +107,21 @@ def __init__(
 
         self.instancenorm = nn.InstanceNorm1d(input_dim)
 
+        if self.use_torch_spec:
+            self.torch_spec = torch.nn.Sequential(
+                PreEmphasis(audio_config["preemphasis"]),
+                torchaudio.transforms.MelSpectrogram(
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/speaker_encoder/models/resnet.py:113]: I tried to do the TorchSTFT compatible with the torchaudio.transforms.MelSpectrogram but have a lot of things differently.  I found and fix a lot of differences but It wasn't enough I think it will take a long time to turn this compatible.  Right now I don't have enough time for this :(. As we are using the checkpoint provided by the authors in the official implementation we need to keep the torchaudio.

File: TTS/tts/utils/speakers.py (hunk 24 above)
Comment (Edresson) [reply TTS/tts/utils/speakers.py]: done

File: TTS/speaker_encoder/models/resnet.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/speaker_encoder/models/resnet.py:181]: I added a comment in the code to turn more clear

File: TTS/tts/models/vits.py (hunk 8 above)
Comment (Edresson) [reply TTS/tts/models/vits.py]: Done

File: TTS/tts/datasets/formatters.py [hunk 37]
```diff
@@ -59,46 +59,56 @@ def mozilla_de(root_path, meta_file):
     return items
 
 
-def mailabs(root_path, meta_files=None):
+def mailabs(root_path, meta_files=None, ununsed_speakers=None):
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/tts/datasets/formatters.py:62]: Done

File: TTS/tts/models/vits.py (hunk 6 above)
Comment (Edresson) [reply TTS/tts/models/vits.py]: Done

File: TTS/tts/utils/synthesis.py (hunk 14 above)
Comment (WeberJulian) [reply TTS/tts/utils/synthesis.py:74]: Done

File: TTS/tts/models/vits.py (hunk 17 above)
Comment (WeberJulian) [reply TTS/tts/models/vits.py:235]: Done

File: TTS/tts/models/vits.py (hunk 20 above)
Comment (WeberJulian) [reply TTS/tts/models/vits.py:568]: Done

File: TTS/tts/models/vits.py (hunk 19 above)
Comment (WeberJulian) [reply TTS/tts/models/vits.py:576]: Should we change to that check for `if self.args.use_speaker_embedding and sid is not None:` and inference ?
Also maybe it's better not to check for `lid is not None`, that way the crash is more explicit (unexpected None instead of a shape error in the encoder) what do you guys think?

File: notebooks/VITS_d-vector_multilingual_exemple.ipynb (hunk 12 above)
Comment (WeberJulian) [reply notebooks/VITS_d-vector_multilingual_exemple.ipynb]: Deleted

File: TTS/speaker_encoder/models/resnet.py (hunk 36 above)
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/speaker_encoder/models/resnet.py]: I removed torchaudio as a requirement. But now it is no longer possible to use the speaker consistency loss when the sample rate of the speaker encoder and the TTS model are different. I left the code that uses torchaudio for resample commented out. This way, if you find an efficient way to resample an audio batch without using torchaudio you can easily implement that.

File: TTS/tts/models/vits.py (hunk 5 above)
Comment (Edresson) [reply TTS/tts/models/vits.py:273]: We need it to know the sample rate of the TTS model to resample the wav or raise an error when the sample rate of the TTS model differs from the speaker encoder sample rate.

File: TTS/tts/datasets/formatters.py [hunk 38]
```diff
@@ -59,46 +59,56 @@ def mozilla_de(root_path, meta_file):
...
         speaker_name = speaker_name_match.group("speaker_name")
+        # ignore speakers
+        if isinstance(ununsed_speakers, list):
+            if speaker_name in ununsed_speakers:
+                continue
         print(" | > {}".format(csv_file))
         with open(txt_file, "r", encoding="utf-8") as ttf:
             for line in ttf:
                 cols = line.split("|")
-                if meta_files is None:
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/tts/datasets/formatters.py:89]: explained above

File: TTS/tts/layers/losses.py [hunk 39]
```diff
@@ -585,6 +586,11 @@ def kl_loss(z_p, logs_q, m_p, logs_p, z_mask):
         l = kl / torch.sum(z_mask)
         return l
 
+    @staticmethod
+    def cosine_similarity_loss(gt_spk_emb, syn_spk_emb):
+        l = - torch.nn.functional.cosine_similarity(gt_spk_emb, syn_spk_emb).mean()
+        return l
+
     def forward(
```
Comment (erogol) [inline TTS/tts/layers/losses.py:594]: Optionally would be nice to have a test for the loss. It is quite a complex loss layer :)

File: TTS/tts/models/base_tts.py [hunk 40]
```diff
@@ -92,7 +93,7 @@ def init_multispeaker(self, config: Coqpit):
                 config.d_vector_dim if "d_vector_dim" in config and config.d_vector_dim is not None else 512
             )
         # init speaker embedding layer
-        if config.use_speaker_embedding and not config.use_d_vector_file:
```
Comment (erogol) [inline TTS/tts/models/base_tts.py:95]: Why did you remove this ?

File: TTS/tts/utils/speakers.py [hunk 41]
```diff
@@ -378,3 +414,13 @@ def get_speaker_manager(c: Coqpit, data: List = None, restore_path: str = None,
...
+
+
+def get_speaker_weighted_sampler(items: list):
+    speaker_names = np.array([item[2] for item in items])
+    unique_speaker_names = np.unique(speaker_names).tolist()
+    speaker_ids = [unique_speaker_names.index(l) for l in speaker_names]
+    speaker_count = np.array([len(np.where(speaker_names == l)[0]) for l in unique_speaker_names])
+    weight_speaker = 1.0 / speaker_count
+    dataset_samples_weight = torch.from_numpy(np.array([weight_speaker[l] for l in speaker_ids])).double()
+    return WeightedRandomSampler(dataset_samples_weight, len(dataset_samples_weight))
```
Comment (erogol) [inline TTS/tts/utils/speakers.py:426]: Tests for the sampler

File: TTS/utils/audio.py [hunk 42]
```diff
@@ -32,6 +32,9 @@ def __init__(
         use_mel=False,
         do_amp_to_db=False,
         spec_gain=1.0,
+        power=None,
+        use_htk=False,
+        mel_norm="slaney"
```
Comment (erogol) [inline TTS/utils/audio.py:37]: Docstring for these ?

What is htk ?

File: TTS/utils/vad.py [hunk 43]
```diff
@@ -0,0 +1,142 @@
+# This code is adpated from: https://github.com/wiseman/py-webrtcvad/blob/master/example.py
```
Comment (erogol) [inline TTS/utils/vad.py:1]: Tests ?

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/bin/remove_silence_using_vad.py [hunk 44]
```diff
@@ -0,0 +1,91 @@
+import os
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/bin/remove_silence_using_vad.py:1]: Done

File: TTS/bin/find_unique_phonemes.py [hunk 45]
```diff
@@ -0,0 +1,61 @@
+"""Find all the unique characters in a dataset"""
+import argparse
```
[the remark this answers is not in the document]
Comment (Edresson) [reply TTS/bin/find_unique_phonemes.py:2]: Done

File: TTS/utils/audio.py (hunk 42 above)
Comment (Edresson) [reply TTS/utils/audio.py:37]: All are Librosa parameters. This class doesn't have a docstring, so I need to create it?

File: TTS/utils/vad.py (hunk 43 above)
Comment (Edresson) [reply TTS/utils/vad.py:1]: It is called by  TTS/bin/remove_silence_using_vad.py so the test of this is not necessary

File: TTS/utils/audio.py (hunk 42 above)
Comment (Edresson) [reply TTS/utils/audio.py:91]: I added a simple docstring, feel free to improve it.

File: TTS/tts/models/base_tts.py (hunk 40 above)
Comment (WeberJulian) [reply TTS/tts/models/base_tts.py:95]: reverted

File: TTS/tts/utils/speakers.py (hunk 41 above)
Comment (WeberJulian) [reply TTS/tts/utils/speakers.py:426]: Done

**Comment by erogol** [comment]
Finally merged :)
